File: leetcode/707.design-linked-list/go/mylinkedlist.go
package mylinklist

type ListNode struct {
	Val  int
	Next *ListNode
}

type MyLinkedList struct {
	Head *ListNode
	Size int
}

func Constructor() MyLinkedList {
	return MyLinkedList{Head: &ListNode{}}
}

func (this *MyLinkedList) Get(index int) int {
	if index < 0 || this.Size <= index {
		return -1
	}

	curr := this.Head.Next

	for range index {
		curr = curr.Next
	}

	return curr.Val
}

func (this *MyLinkedList) AddAtHead(val int) {
	newHead := ListNode{Val: val, Next: this.Head.Next}
	this.Head.Next = &newHead
	this.Size++
}

func (this *MyLinkedList) AddAtTail(val int) {
	newHead := ListNode{Val: val}

	curr := this.Head

	for curr.Next != nil {
		curr = curr.Next
	}

	curr.Next = &newHead
	this.Size++
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index > this.Size {
		return
	}

	if index < 0 {
		index = 0
	}

	curr := this.Head

	for range index {
		curr = curr.Next
	}

	newHead := ListNode{Val: val, Next: curr.Next}
	curr.Next = &newHead

	this.Size++
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if index < 0 || index >= this.Size {
		return
	}

	curr := this.Head

	for range index {
		curr = curr.Next
	}

	curr.Next = curr.Next.Next
	this.Size--
}

File: leetcode/707.design-linked-list/ts/mylinkedlist.ts
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

export class MyLinkedList {
    head: ListNode
    size: number

    constructor() {
        this.head = new ListNode()
        this.size = 0
    }

    get(index: number): number {
        if (index < 0 || index >= this.size) return -1

        let curr = this.head.next

        for (let i = 0; i < index; i++)
            curr = curr!.next

        return curr!.val
    }

    addAtHead(val: number): void {
        this.head.next = new ListNode(val, this.head.next)
        this.size++
    }

    addAtTail(val: number): void {
        let curr = this.head
        while (curr.next !== null)
            curr = curr.next

        curr.next = new ListNode(val)
        this.size++
    }

    addAtIndex(index: number, val: number): void {
        if (index > this.size) return
        if (index < 0) index = 0

        let curr: ListNode | null = this.head
        for (let i = 0; i < index; i++)
            curr = curr!.next

        curr!.next = new ListNode(val, curr?.next)
        this.size++
    }

    deleteAtIndex(index: number): void {
        if (index < 0 || index >= this.size) return

        let curr: ListNode | null = this.head
        for (let i = 0; i < index; i++)
            curr = curr!.next

        curr!.next = curr!.next!.next
        this.size--
    }
}
